fms_extras/utils/cache/__init__.py

```diff
@@ -0,0 +1,330 @@
+        sequence_ids: List[int]
+            list of sequence ids to free
+        """
+        pass
+
+
+KVCache = Tuple[torch.Tensor, torch.Tensor]  # (key cache, value cache)
+
+
+@dataclasses.dataclass
+class OutOfPlaceCacheDataLayer(CacheDataLayer):
+    data_layer: Tuple[torch.Tensor, torch.Tensor]
+
+    def get_cache_type(self) -> str:
+        return "out-of-place"
+
+    def store(
+        self, keys: torch.Tensor, values: torch.Tensor
+    ) -> Tuple[torch.Tensor, torch.Tensor]:
+        if self.data_layer is not None:
+            self.data_layer = (
+                torch.cat((self.data_layer[0], keys), dim=2),
+                torch.cat((self.data_layer[1], values), dim=2),
+            )
+            keys, values = self.data_layer
+        return keys, values
+
+    def is_filled(self) -> bool:
+        return self.data_layer is not None
+
+
+class OutOfPlaceCacheData(CacheData):
+    def __init__(self, data: List[Tuple[torch.Tensor, torch.Tensor]]):
+        self.data = data
+        self.max_sequence_length = (
+            0 if self.data[0] is None else self.data[0][0].size(2)
+        )
+
+    def get_layer(self, layer_index: int) -> OutOfPlaceCacheDataLayer:
+        return OutOfPlaceCacheDataLayer(data_layer=self.data[layer_index])
+
+    def is_filled(self) -> bool:
+        return self.data[0] is not None
+
+
+def flatten_batch(inp):
+    # Takes a bsize x n_candidates x candidate_len rectangular batch of input indices
+    # Returns 1) a flattened set of indices with all redundant tokens removed,
+    # and 2) a tensor, sized as input, mapping each input token to its slot in output,
+    # and 3) a tensor, sized as output, mapping each output token to slot in flattened input
+    ind_out = torch.zeros_like(inp)
+    inp = inp.tolist()
+    out = []
+    ind_flat = []
+    batch_offset = 0
+    for b, candidate_set in enumerate(inp):
+        lineages = []
+        for k, candidate in enumerate(candidate_set):
+            for n in range(len(candidate)):
+                lineage = tuple(candidate[: n + 1])
+                if lineage in lineages:
+                    # Token is redundant
+                    ind_out[b, k, n] = lineages.index(lineage) + batch_offset
+                else:
+                    # Token is not redundant
+                    ind_out[b, k, n] = len(lineages) + batch_offset
+                    lineages.append(lineage)
+                    ind_flat.append(
+                        b * len(inp[0]) * len(inp[0][0]) + k * len(inp[0][0]) + n
+                    )
+        out.append(
+            torch.tensor(
+                [lineage[-1] for lineage in lineages],
+                device=ind_out.device,
+                dtype=torch.int32,
+            )
+        )
+        batch_offset += len(lineages)
+    return (
+        torch.cat(out),
+        ind_out,
+        torch.tensor(ind_flat, device=ind_out.device, dtype=torch.int32),
+    )
+
+
+def select_inflate_dim(inp, inds, dim=0):
```

Comment:
what do the dimension letters mean here?

Reply:
This is a batched index select that respects the shape of the index tensor. So if index tensor is size (a, ..., z) and dim=0, we'll select the index values from input axis 0, and return the contents in shape (a, ..., z, ...) with trailing dimensions coming from the trailing dimensions of the input tensor. Like above, it probably does belong as a speculator utility in fms-extras.